// @pushchain/core/src/lib/constants/abi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC20_EVM = exports.UNIVERSAL_GATEWAY_V0 = exports.UEA_SVM = exports.UEA_EVM = exports.FACTORY_V1 = exports.FEE_LOCKER_SVM = exports.FEE_LOCKER_EVM = void 0;
const tslib_1 = require("tslib");
var feeLocker_evm_1 = require("./feeLocker.evm");
Object.defineProperty(exports, "FEE_LOCKER_EVM", { enumerable: true, get: function () { return feeLocker_evm_1.FEE_LOCKER_EVM; } });
var feeLocker_json_1 = require("./feeLocker.json");
Object.defineProperty(exports, "FEE_LOCKER_SVM", { enumerable: true, get: function () { return tslib_1.__importDefault(feeLocker_json_1).default; } });
var factoryV1_1 = require("./factoryV1");
Object.defineProperty(exports, "FACTORY_V1", { enumerable: true, get: function () { return factoryV1_1.FACTORY_V1; } });
var uea_evm_1 = require("./uea.evm");
Object.defineProperty(exports, "UEA_EVM", { enumerable: true, get: function () { return uea_evm_1.UEA_EVM; } });
var uea_svm_1 = require("./uea.svm");
Object.defineProperty(exports, "UEA_SVM", { enumerable: true, get: function () { return uea_svm_1.UEA_SVM; } });
var universalGatewayV0_evm_1 = require("./universalGatewayV0.evm");
Object.defineProperty(exports, "UNIVERSAL_GATEWAY_V0", { enumerable: true, get: function () { return universalGatewayV0_evm_1.UNIVERSAL_GATEWAY_V0; } });
var erc20_evm_1 = require("./erc20.evm");
Object.defineProperty(exports, "ERC20_EVM", { enumerable: true, get: function () { return erc20_evm_1.ERC20_EVM; } });
