// @pushchain/core/src/lib/constants/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LIBRARY = exports.VM = exports.CHAIN = exports.PUSH_NETWORK = void 0;
/**
 * Specifies the Push Chain environment the SDK will connect to.
 * Determines the RPC endpoints, contract addresses, and network context
 * for interacting with Push Chain.
 */
var PUSH_NETWORK;
(function (PUSH_NETWORK) {
    PUSH_NETWORK["MAINNET"] = "MAINNET";
    PUSH_NETWORK["TESTNET_DONUT"] = "TESTNET_DONUT";
    PUSH_NETWORK["TESTNET"] = "TESTNET";
    PUSH_NETWORK["LOCALNET"] = "LOCALNET";
})(PUSH_NETWORK || (exports.PUSH_NETWORK = PUSH_NETWORK = {}));
/**
 * Supported Chains
 */
var CHAIN;
(function (CHAIN) {
    // Push
    CHAIN["PUSH_MAINNET"] = "eip155:9";
    CHAIN["PUSH_TESTNET"] = "eip155:42101";
    CHAIN["PUSH_TESTNET_DONUT"] = "eip155:42101";
    CHAIN["PUSH_LOCALNET"] = "eip155:9001";
    // Ethereum
    CHAIN["ETHEREUM_MAINNET"] = "eip155:1";
    CHAIN["ETHEREUM_SEPOLIA"] = "eip155:11155111";
    // Solana
    CHAIN["SOLANA_MAINNET"] = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
    CHAIN["SOLANA_TESTNET"] = "solana:4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z";
    CHAIN["SOLANA_DEVNET"] = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
})(CHAIN || (exports.CHAIN = CHAIN = {}));
/**
 * Represents the virtual machine or execution environment for a chain.
 */
var VM;
(function (VM) {
    VM["EVM"] = "EVM";
    VM["SVM"] = "SVM";
})(VM || (exports.VM = VM = {}));
var LIBRARY;
(function (LIBRARY) {
    LIBRARY["ETHEREUM_VIEM"] = "viem";
    LIBRARY["SOLANA_WEB3JS"] = "solana-web3js";
    LIBRARY["ETHEREUM_ETHERSV6"] = "ethers-v6";
})(LIBRARY || (exports.LIBRARY = LIBRARY = {}));
