// @pushchain/core/src/lib/progress-hook/progress-hook.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROGRESS_HOOK = void 0;
var PROGRESS_HOOK;
(function (PROGRESS_HOOK) {
    PROGRESS_HOOK["SEND_TX_01"] = "SEND-TX-01";
    PROGRESS_HOOK["SEND_TX_02_01"] = "SEND-TX-02-01";
    PROGRESS_HOOK["SEND_TX_02_02"] = "SEND-TX-02-02";
    PROGRESS_HOOK["SEND_TX_03_01"] = "SEND-TX-03-01";
    PROGRESS_HOOK["SEND_TX_03_02"] = "SEND-TX-03-02";
    PROGRESS_HOOK["SEND_TX_04_01"] = "SEND-TX-04-01";
    PROGRESS_HOOK["SEND_TX_04_02"] = "SEND-TX-04-02";
    // V2 funds flow additions
    PROGRESS_HOOK["SEND_TX_04_03"] = "SEND-TX-04-03";
    PROGRESS_HOOK["SEND_TX_04_04"] = "SEND-TX-04-04";
    PROGRESS_HOOK["SEND_TX_05_01"] = "SEND-TX-05-01";
    PROGRESS_HOOK["SEND_TX_05_02"] = "SEND-TX-05-02";
    PROGRESS_HOOK["SEND_TX_05_03"] = "SEND-TX-05-03";
    // Broadcasting to Push Chain (existing)
    PROGRESS_HOOK["SEND_TX_06"] = "SEND-TX-06";
    // Funds flow granularity (origin chain)
    PROGRESS_HOOK["SEND_TX_06_01"] = "SEND-TX-06-01";
    PROGRESS_HOOK["SEND_TX_06_02"] = "SEND-TX-06-02";
    PROGRESS_HOOK["SEND_TX_06_03"] = "SEND-TX-06-03";
    PROGRESS_HOOK["SEND_TX_06_04"] = "SEND-TX-06-04";
    PROGRESS_HOOK["SEND_TX_06_05"] = "SEND-TX-06-05";
    PROGRESS_HOOK["SEND_TX_06_06"] = "SEND-TX-06-06";
    PROGRESS_HOOK["SEND_TX_06_07"] = "SEND-TX-06-07";
    PROGRESS_HOOK["SEND_TX_99_01"] = "SEND-TX-99-01";
    PROGRESS_HOOK["SEND_TX_99_02"] = "SEND-TX-99-02";
})(PROGRESS_HOOK || (exports.PROGRESS_HOOK = PROGRESS_HOOK = {}));
